File: esp32/plugins/telegraf.py
from mqtt_connection import ErikaMqqt
from plugins.erika_plugin_base import ErikaBasePlugin
import json

class Telegraf(ErikaBasePlugin):
    def __init__(self, erika:Erika=None, erika_mqqt:ErikaMqqt=None): # type: ignore
        
        super().__init__(
            erika=erika, 
            erika_mqqt=erika_mqqt,
            keylogging=True,
            #active=True
            )
    def on_message(self, topic: str, msg: str):
        return super().on_message(topic, "Telegraf received: "+msg)

    async def on_keystroke(self, key=""):
        erika_mqqt = self.erika_mqqt
        data={"sender": self.erika.uuid, "text":key}
        await erika_mqqt.client.publish(erika_mqqt.channel_print_all, json.dumps(data), qos=0)
        return

File: virtual_erika/plugins/telegraf.js
// Telegraf Plugin for Virtual Erika
// Mirrors the Python telegraf.py plugin functionality

class Telegraf extends VirtualErikaBasePlugin {
    constructor(virtualErika) {
        super(virtualErika, {
            topic: 'telegraf',
            keylogging: true,
            active: false, // Start inactive
            info: "Ein Telegraf, der an alle Erikas sendet, die gerade an sind. ON/OFF"
        });
    }

    onMessage(topic, msg) {
        if (this.active) {
            this.virtualErika.writeToOutput(`Telegraf received: ${msg}`, 'system');
            this.virtualErika.writeToIO(`Telegraf: ${msg}`, 'output');
        }
    }

    onKeystroke(key) {
        console.log(`Telegraf plugin onKeystroke called with key: ${key}, active: ${this.active}, keylogging: ${this.keylogging}`);
        
        if (this.active && this.keylogging) {
            // Send keystroke to all Erikas via print_all channel (like real Telegraf)
            const data = {
                sender: this.virtualErika.config.DEVICE_UUID,
                text: key
            };
            
            this.sendMessage(this.virtualErika.config.TOPICS.PRINT_ALL, data);
            this.virtualErika.writeToOutput(`Telegraf broadcast: ${key}`, 'system');
        } else {
            console.log(`Telegraf plugin not active or not keylogging`);
        }
    }
} 